src/AdoNet/Orleans.Persistence.AdoNet/Options/AdoNetGrainStorageOptions.cs

```diff
@@ -0,0 +1,138 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+using Microsoft.Extensions.Options;
+using Newtonsoft.Json;
+using Orleans.Persistence.AdoNet.Storage;
+using Orleans.Runtime;
+using Orleans.Runtime.Configuration;
+
+namespace Orleans.Hosting
+{
+    /// <summary>
+    /// Options for AdonetGrainStorage
+    /// </summary>
+    public class AdoNetGrainStorageOptions
+    {
+        /// <summary>
+        /// Connection string for AdoNet storage.
+        /// </summary>
+        public string ConnectionString { get; set; }
+
+        /// <summary>
+        /// Stage of silo lifecycle where storage should be initialized.  Storage must be initialzed prior to use.
+        /// </summary>
+        public int InitStage { get; set; } = DEFAULT_INIT_STAGE;
```

Comment:
I think this shouldn't be configurable

Reply:
My reasoning:
* Other lifecycle participants must be able to rely on certain things happening at certain stages - that's the point of having named stages
* There's no good reason for users to change this value

Reply:
Yeah this is definitely an advanced feature. Maybe we shouldn't make it this discoverable. @jason-bragg on what's his thought on this. He added this when he is doing the AzureTableGrainStorage. He may have his reasoning. 

But either way, I think we can merge this first. It is accumulating merge conflicts everyday:) 

If we decided that this is not a good idea, we can submit another pr to fix it from all GrainStorageProviders.

Reply:
The reason for allowing users to configure when grain storage starts is because some systems, like versioning, start very early in the lifecycle, and can be configured to use storage, while some storage systems themselves use grains which may trigger systems that use storage (leasing or transactions for instance.  Dependencies between systems are not always clear.  Allowing this to be configurable, along with the wide ranges between the predefined stages, give application developers a way out should they use a component that has unexpected dependency chains.  As with many of our settings, the default should be used unless their is reasons to change, so most users will never need to concern themselves with this, but for the cases where they need to change this, they can.

Reply:
I see your point. I'm concerned that breaking basic assumptions about initialization order introduces too much complexity. Debugging issues surrounding it would be difficult.

If it's fine for storage providers to sometimes start earlier, then we should always start them earlier.

Reply:
Since we're porting these from providers, and the original logic expected initialization at startup, I've advocated preserving the pattern of initializing on startup.  Since we're initializing on startup, we're taking part in the lifecycle and should probably afford users the ability to specify the stage for dependency issues.

Having said all that, in the future I'd prefer to use the lifecycle primarily for starting systems that are services (like the stream providers) or have heavy initialization (like transaction manager) rather than logic which just needs asynchronous initialization (like storage provider).  For things that simply require asynchronous initialization, I'd prefer we perform a lazy initialization, on demand, rather than plugging them into the lifecycle.  Barring circular dependencies (which can happen) lazy initialization is a simpler patter.  Basically if a system has no other dependencies, it probably shouldn't be initialized via the lifecycle, it can, but probably shouldn't.

These initial ports away from the provider model open the door for many improvements to our storage patterns, and I don't suspect that most should really need to take part in the lifecycle once fully fitted to the new patterns.  I suspect we can/will clean this up once we move grain storage to facets.  Since the public IGrainStorage interface is now independent of the initialization, this all becomes implementation details which only effect how a provider is configured.

In short, we can do better, this is just once step.